src/main/java/core/basesyntax/Car.java

```diff
@@ -1,13 +1,86 @@
 package core.basesyntax;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 
-/**
- * Make this class immutable. See requirements in the README file
- */
-public class Car {
-    private int year;
-    private String color;
-    private List<Wheel> wheels;
-    private Engine engine;
+public final class Car implements Cloneable {
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = cloneList(wheels);
+        this.engine = engine == null ? null : engine.clone();
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(year, color, wheels, engine);
+    }
+
+    public Car changeColor(String newColor) {
+        return new Car(year, newColor, wheels, engine);
+    }
+
+    public Car addWheel(Wheel wheel) {
+        List<Wheel> newList = cloneList(wheels);
+        newList.add(wheel);
+        return new Car(year, color, newList, engine);
+    }
+
+    public int getYear() {
+        return year;
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    public List<Wheel> getWheels() {
+        return cloneList(wheels);
+    }
+
+    public Engine getEngine() {
+        return engine == null ? null : engine.clone();
+    }
+
+    private List<Wheel> cloneList(List<Wheel> wheels) {
+        List<Wheel> clonedList = new ArrayList<>();
```

Comment:
```suggestion
        List<Wheel> clonedList = new ArrayList<>(wheels.size());
```
you can pass capacity value and if it's more than 10 it will benefit performance